lib/cusolver/cusolverRF.jl

```diff
@@ -0,0 +1,429 @@
+    cusolverSpDcsrluBufferInfoHost(
+        spH,
+        n, nnzA, desca,
+        h_valsB, h_rowsB, h_colsB,
+        info[],
+        size_internal, size_lu
+    )
+
+    n_bytes = size_lu[] * sizeof(Cint)
+    buffer_lu = Base.Libc.malloc(n_bytes)
+    pivot_threshold = 1.0
+
+    cusolverSpDcsrluFactorHost(
+        spH, n, nnzA, desca,
+        h_valsB, h_rowsB, h_colsB,
+        info[], pivot_threshold,
+        buffer_lu,
+    )
+
+    # Check singularity
+    if check
+        singularity = Ref{Cint}(0)
+        cusolverSpDcsrluZeroPivotHost(
+            spH, info[], tol, singularity,
+        )
+
+        # Check that the matrix is nonsingular
+        if singularity[] >= 0
+            SingularException(singularity[])
+        end
+    end
+
+    # Get size of L and U
+    pnnzU = Ref{Cint}(0)
+    pnnzL = Ref{Cint}(0)
+    cusolverSpXcsrluNnzHost(
+        spH,
+        pnnzL, pnnzU, info[],
+    )
+
+    nnzL = pnnzL[]
+    nnzU = pnnzU[]
+
+    # Retrieve L and U matrices
+    h_Plu = zeros(Cint, m)
+    h_Qlu = zeros(Cint, n)
+
+    h_valsL = zeros(nnzL)
+    h_rowsL = zeros(Cint, m+1)
+    h_colsL = zeros(Cint, nnzL)
+
+    h_valsU = zeros(nnzU)
+    h_rowsU = zeros(Cint, m+1)
+    h_colsU = zeros(Cint, nnzU)
+
+    # Extract
+    cusolverSpDcsrluExtractHost(
+        spH,
+        h_Plu, h_Qlu,
+        desca,
+        h_valsL, h_rowsL, h_colsL,
+        desca,
+        h_valsU, h_rowsU, h_colsU,
+        info[],
+        buffer_lu,
+    )
+
+    h_P = h_Qreorder[h_Plu .+ 1]
+    h_Q = h_Qreorder[h_Qlu .+ 1]
+
+    if buffer_lu != C_NULL
+        Base.Libc.free(buffer_lu)
+        buffer_lu = C_NULL
+    end
+    if buffer_cpu != C_NULL
+        Base.Libc.free(buffer_cpu)
+        buffer_cpu = C_NULL
+    end
+
+    return RfHostLU(
+        nnzA, h_rowsA, h_colsA, h_valsA,
+        nnzL, h_rowsL, h_colsL, h_valsL,
+        nnzU, h_rowsU, h_colsU, h_valsU,
+        h_P, h_Q,
+    )
+end
+
+struct RfLU{T} <: LinearAlgebra.Factorization{T}
+    rf::RfHandle
+    nrhs::Int
+    n::Int
+    m::Int
+    nnzA::Int
+    drowsA::CuVector{Cint}
+    dcolsA::CuVector{Cint}
+    dP::CuVector{Cint}
+    dQ::CuVector{Cint}
+    dT::CuVector{T}
+end
+
+function RfLU(
+    A::CuSparseMatrixCSR{T};
+    nrhs=1, fast_mode=true, ordering=:AMD,
+) where T
+    if nrhs > 1
+        error("Currently CusolverRF supports only one right-hand side.")
+    end
+    n, m = size(A)
+    lu_host = RfHostLU(A; ordering=ordering)
+
+    # Allocations (device)
+    d_T = CUDA.zeros(Cdouble, m * nrhs)
+
+    rf = RfHandle(; fast_mode=fast_mode)
+
+    # Assemble internal data structures
+    cusolverRfSetupHost(
+        n, lu_host.nnzA, lu_host.rowsA, lu_host.colsA, lu_host.valsA,
+        lu_host.nnzL, lu_host.rowsL, lu_host.colsL, lu_host.valsL,
+        lu_host.nnzU, lu_host.rowsU, lu_host.colsU, lu_host.valsU,
+        lu_host.P, lu_host.Q,
+        rf
+    )
+    # Analyze available parallelism
+    CUDA.@sync cusolverRfAnalyze(rf)
```

Comment:
Why does this need to synchronize?

Reply:
Good question. I was not sure whether to add `@sync` here or not. Furthermore, the documentation of cusolverRF is not clear about that. I have chosen to include them as: 
- on some large instances, removing `@sync` leads to wrong results. 
- in the [example provided in the documentation](https://docs.nvidia.com/cuda/cusolver/index.html#cuSolverRFbatch-example1), they add explicit `cudaDeviceSynchronize()` after `cusolverRfAnalyse`, `cusolverRfBatchRefactor` and `cusolverRfBatchSolve`. 

But I may have missed something.